### PR TITLE
[processor/lsminterval] Add support for exponential histogram

### PR DESCRIPTION
**Pull request by lahsivjar**
Merges delta histogram with downscaling if required to merge histograms with 2 different scales.

Closes: https://github.com/elastic/opentelemetry-collector-components/issues/126

### CODE REVIEW

File: processor/lsmintervalprocessor/internal/merger/model.go
Comment (lahsivjar) [inline processor/lsmintervalprocessor/internal/merger/model.go:439]: [**For reviewers**] Will do this in a followup.

File: processor/lsmintervalprocessor/internal/merger/model.go
Comment (lahsivjar) [inline processor/lsmintervalprocessor/internal/merger/model.go:500]: [**For reviewers**] I don't think this is a big issue (keeping trailing zeroes) but let me know if there are any concerns.

File: processor/lsmintervalprocessor/internal/merger/model.go
Comment (lahsivjar) [inline processor/lsmintervalprocessor/internal/merger/model.go:422]: We should short-circuit these 2 conditions by just copying the whole DO. I haven't done this because I wanted to confirm if zero count is included in count but then it slipped my mind. I will fix this

File: processor/lsmintervalprocessor/internal/merger/model.go
Comment (axw) [reply processor/lsmintervalprocessor/internal/merger/model.go:439]: FYI there's logic you could copy at https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/processor/deltatocumulativeprocessor/internal/data/expo/zero.go

I wonder if we should just vendor and use that code, under the assumption that lsmintervalprocessor can be moved to opentelemetry-collector-contrib. At that point we could create a common (internal?) package for merging exponential histograms -- or maybe metrics in general.

File: processor/lsmintervalprocessor/internal/merger/model.go
Comment (axw) [inline processor/lsmintervalprocessor/internal/merger/model.go:541]: Nit: use `cmp.Ordered` and `cmp.Less` instead of the experimental constraints package?

File: processor/lsmintervalprocessor/internal/merger/model.go
Comment (axw) [reply processor/lsmintervalprocessor/internal/merger/model.go:499]: I wouldn't worry about it for now.

**Review by axw — COMMENTED** [review]
Before I dive into the algorithms: is there a good reason why we shouldn't just vendor/copy a bit of code from deltatocumulativeprocessor temporarily? If not, I'd prefer we do that.

File: processor/lsmintervalprocessor/internal/merger/model.go
Comment (1pkg) [reply processor/lsmintervalprocessor/internal/merger/model.go:541]: Better yet use min and max builtin from the standard library https://pkg.go.dev/builtin#min, they are available since 1.21

Comment (axw) [reply processor/lsmintervalprocessor/internal/merger/model.go:541]: Yes! I thought there were functions, just couldn't remember where they were.

File: processor/lsmintervalprocessor/internal/merger/model.go
Comment (lahsivjar) [reply processor/lsmintervalprocessor/internal/merger/model.go:439]: > FYI there's logic you could copy

Ah, thanks! This would be helpful.

> I wonder if we should just vendor and use that code

I don't think we can. Looked into the code briefly and it only merges buckets of same scale: https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/a1915508175a498b731d17c951d776761ce8e824/processor/deltatocumulativeprocessor/internal/data/expo/merge.go#L10-L12

Maybe I will try to upstream the changes in this PR to support merges across scale.

File: processor/lsmintervalprocessor/internal/merger/model.go
Comment (lahsivjar) [reply processor/lsmintervalprocessor/internal/merger/model.go:541]: > Better yet use min and max builtin from the standard library https://pkg.go.dev/builtin#min, they are available since 1.21

wth, I spent time trying to find these and failed 🤦

Comment (lahsivjar) [reply processor/lsmintervalprocessor/internal/merger/model.go]: Fixed, thanks for this!! I don't know why I remembered it being in some other package and not in standard library 🤦

File: processor/lsmintervalprocessor/internal/merger/model.go
Comment (lahsivjar) [reply processor/lsmintervalprocessor/internal/merger/model.go:424]: This is done!

**Comment by lahsivjar** [comment]
> Before I dive into the algorithms: is there a good reason why we shouldn't just vendor/copy a bit of code from deltatocumulativeprocessor temporarily? If not, I'd prefer we do that.

@axw Replied at https://github.com/elastic/opentelemetry-collector-components/pull/151#discussion_r1783988351. TL;DR, that one only supports merges for same scale.

File: processor/lsmintervalprocessor/internal/merger/model.go
Comment (lahsivjar) [reply processor/lsmintervalprocessor/internal/merger/model.go:507]: 👍 For posterity (and for me as well when I read this back), this is easy to do by just doing right-shift on the effective len and reslicing. The only reason to not do is due to the extra reallocation involved. Sample code:

```
	end := (hist.BucketCounts().Len() + int(hist.Offset()) - 1) >> scaleDiff // using offset before scaledown
	hist.BucketCounts().FromRaw(hist.BucketCounts().AsRaw()[:(end + 1)])
```

File: processor/lsmintervalprocessor/internal/merger/model.go
Comment (lahsivjar) [reply processor/lsmintervalprocessor/internal/merger/model.go:447]: > Maybe I will try to upstream the changes in this PR to support merges across scale.

Just realized that it doesn't make sense for deltatocumulative to support multiple scales. Maybe we can create a common exphistogram package in future when we try to upstream this.

Comment (axw) [reply processor/lsmintervalprocessor/internal/merger/model.go:447]: @lahsivjar it's true that the Merge function only works with buckets of the same scale, but there's also code to downscale: https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/a1915508175a498b731d17c951d776761ce8e824/processor/deltatocumulativeprocessor/internal/data/expo/scale.go#L42

They're all put together here: https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/a1915508175a498b731d17c951d776761ce8e824/processor/deltatocumulativeprocessor/internal/data/add.go#L68-L110

Comment (lahsivjar) [reply processor/lsmintervalprocessor/internal/merger/model.go:447]: Hmm, the logic is very similar but I am a bit reluctant since we will have to vendor the code and build `Merge` on top of it.

Comment (lahsivjar) [reply processor/lsmintervalprocessor/internal/merger/model.go:447]: I think the collapse logic is a bit complicated upstream in terms of how it is written - nothing major. It seems I am getting attached to my code 😅 . It might be better to just vendor. Let me update!

File: processor/lsmintervalprocessor/internal/merger/model.go
Comment (carsonip) [inline processor/lsmintervalprocessor/internal/merger/model.go:473]: how will this work when both to and from offsets are negative?

File: processor/lsmintervalprocessor/internal/merger/model.go
Comment (carsonip) [inline processor/lsmintervalprocessor/internal/merger/model.go:467]: Q: I'm slightly confused by this line. Is the meaning of "size" here the same as fromSize and toSize, or does it mean bucket length? the most intuitive way to me to get the new number of buckets is max(fromBuckets.Len(),toBuckets.Len()) + abs(from.Offset()-to.Offset()). Is this the same?

File: processor/lsmintervalprocessor/internal/merger/model.go
Comment (lahsivjar) [reply processor/lsmintervalprocessor/internal/merger/model.go:473]: I don't think offsets can be negative, they are number of zeroes and negative buckets are separated. Anway, check this https://github.com/elastic/opentelemetry-collector-components/pull/151#discussion_r1783675061 (TL;DR, we are vendoring an upstream internal implementation)

File: processor/lsmintervalprocessor/internal/merger/model.go
Comment (lahsivjar) [reply processor/lsmintervalprocessor/internal/merger/model.go:467]: Size here is the bucket size after merging. If you include the offset then you can calculate the size using above formula.

File: processor/lsmintervalprocessor/internal/merger/model.go
Comment (lahsivjar) [reply processor/lsmintervalprocessor/internal/merger/model.go]: Updated the code to vendor the internal package. Will also update histogram to use that instead in a separate PR.

File: processor/lsmintervalprocessor/internal/merger/model.go
Comment (lahsivjar) [reply processor/lsmintervalprocessor/internal/merger/model.go]: For posterity, Carson is correct here. The offset cannot be taken directly to calculate indexes, they must be normalized first by taking the difference of max of the upper bounds and min of the lower bounds.

File: processor/lsmintervalprocessor/internal/data/datatest/equal_test.go
Comment (lahsivjar) [inline processor/lsmintervalprocessor/internal/data/datatest/equal_test.go:63]: [**For reviewers**] This bit is changed from upstream (line number from 35 -> 55) because the test is a bit brittle. It parses these lines (which has line numbers) and compares it with the actual output 😞

File: processor/lsmintervalprocessor/internal/merger/model_test.go
Comment (axw) [inline processor/lsmintervalprocessor/internal/merger/model_test.go:52]: Do we need all of these test cases here? I would think we can defer to the testing of package expo, and just verify here that things are wired up correctly with a basic test.

**Review by axw — APPROVED** [review]
LGTM, thanks for the changes :)

File: processor/lsmintervalprocessor/internal/merger/model_test.go
Comment (lahsivjar) [reply processor/lsmintervalprocessor/internal/merger/model_test.go:52]: Yup, sound good. Let me remove that!

Comment (lahsivjar) [reply processor/lsmintervalprocessor/internal/merger/model_test.go]: Done!

**Pull request merged**
